src/gp_init.c

```diff
@@ -376,47 +376,101 @@ int init_event_fini(struct gssproxy_ctx *gpctx)
     return 0;
 }
 
-void init_proc_nfsd(struct gp_config *cfg)
+static int init_proc_nfsd_once(void)
 {
     char buf[] = "1";
-    bool enabled = false;
     int fd, ret;
-    static int poked = 0;
-
-    /* check first if any service enabled kernel support */
-    for (int i = 0; i < cfg->num_svcs; i++) {
-        if (cfg->svcs[i]->kernel_nfsd) {
-            enabled = true;
-            break;
-        }
-    }
+    static bool poked = false;
+    static bool warned_once = false;
 
-    if (!enabled || poked) {
-        return;
-    }
+    if (poked)
+        return 0;
 
     fd = open(LINUX_PROC_USE_GSS_PROXY_FILE, O_RDWR);
     if (fd == -1) {
         ret = errno;
-        GPDEBUG("Kernel doesn't support GSS-Proxy (can't open %s: %d (%s))\n",
-                LINUX_PROC_USE_GSS_PROXY_FILE, ret, gp_strerror(ret));
-        goto fail;
+        if (!warned_once) {
+            GPDEBUG("Kernel doesn't support GSS-Proxy "
+                    "(can't open %s: %d (%s))\n",
+                    LINUX_PROC_USE_GSS_PROXY_FILE, ret, gp_strerror(ret));
+            warned_once = true;
+        }
+        goto out;
     }
 
     ret = write(fd, buf, 1);
     if (ret != 1) {
         ret = errno;
         GPDEBUG("Failed to write to %s: %d (%s)\n",
                 LINUX_PROC_USE_GSS_PROXY_FILE, ret, gp_strerror(ret));
-        close(fd);
-        goto fail;
+        goto out_close;
     }
 
-    poked = 1;
+    GPDEBUG("Kernel GSS-Proxy support enabled\n");
+    poked = true;
+    ret = 0;
+out_close:
```

Comment:
please let's not use multiple out conditions.
Rather initialize fd to -1, and always got close, int here check if fd != -1 then  close(fd)

Reply:
Or just `close(fd)` since the return value of close isn't checked?

Reply:
as log as we always initialize fd to -1 in the function so there is never a random value or 0 that would be fine.